### PR TITLE
First MR to implement merge of parquet file on hdfs reader

### PR DESCRIPTION
**Pull request by ashangit**
Write offset for each messages on each parquet table
Add index on filename

Idea is then to be able from index to get last written files per day and then from parquet metadata to get max offset.

### CODE REVIEW

File: readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputerTest.java [hunk 1]
```diff
@@ -22,11 +22,23 @@
 import static org.mockito.Mockito.when;
 
 public class HdfsOffsetComputerTest {
+    final HdfsOffsetComputer offsetComputer = new HdfsOffsetComputer(buildFileSystem(
```
Comment (gray-matter) [inline readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputerTest.java]: I dislike instance or class variables for tests. I would leave this where it was to ensure that there's no poisoning between tests

File: readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputerTest.java [hunk 2]
```diff
@@ -114,11 +132,11 @@ public void actualFileSystem() throws IOException {
              */
             localFs.mkdirs(rootPath);
             localFs.mkdirs(basePath);
-            localFs.create(new Path(basePath, hdfsOffsetComputer.computePath(today, buildOffset(1, 1))));
-            localFs.create(new Path(basePath, hdfsOffsetComputer.computePath(today, buildOffset(2, 12))));
-            localFs.create(new Path(basePath, hdfsOffsetComputer.computePath(yesterday, buildOffset(1, 2))));
-            localFs.create(new Path(basePath, hdfsOffsetComputer.computePath(yesterday, buildOffset(1, 3))));
-            localFs.create(new Path(basePath, hdfsOffsetComputer.computePath(twoDaysAgo, buildOffset(1, 42))));
+            localFs.create(new Path(basePath, hdfsOffsetComputer.computePath(today, 0L, buildOffset(1, 1))));
```
Comment (gray-matter) [inline readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputerTest.java]: Need to update the comment above, I guess (adding index=0)?

File: readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriterTest.java [hunk 3]
```diff
@@ -241,13 +244,13 @@ public void getStartingOffsets() throws IOException {
...
-        final PartitionedWriter<String> writer = new PartitionedWriter<>((ignored) -> consumer,
-                mock(OffsetComputer.class), "ignored");
+        final ExpiringConsumer<Message> consumer = mock(ExpiringConsumer.class);
+        final PartitionedWriter<Message> writer = new PartitionedWriter<>((ignored) -> consumer,
+                mock(OffsetComputer.class), "ignored", DataAccessEventProtos.FsEvent.newBuilder().build());
 
         writer.heartbeat(partition, offset);
 
-        verify(consumer, times(1)).write(eq(null), any(Offset.class));
+        verify(consumer, times(1)).write(any(Message.class), any(Offset.class));
```
Comment (gray-matter) [inline readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/writer/PartitionedWriterTest.java]: Could check that it's an empty message

File: readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/writer/ProtoParquetWriterWithOffsetTest.java [hunk 4]
```diff
@@ -59,10 +56,12 @@ public void closeWithSomeEvents() throws IOException {
         when(fsMock.mkdirs(any(Path.class))).thenReturn(true);
 
         when(fsMock.rename(any(Path.class), any(Path.class))).thenReturn(true);
+        when(fsMock.globStatus(any(Path.class))).thenReturn(new FileStatus[]{});
```
Comment (gray-matter) [inline readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/writer/ProtoParquetWriterWithOffsetTest.java]: Missing a test case with an existing index file?

File: readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputer.java [hunk 5]
```diff
@@ -45,11 +45,12 @@ public HdfsOffsetComputer(FileSystem fs, Path basePath, @Nullable String kafkaCl
         this.dirRenamePattern = "%s";
 
         if (kafkaCluster == null) {
```
[the remark this answers is not in the document]
Comment (mchataigner) [reply readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputer.java]: nop, we are only using kafkaCluster in prod PA4, nothing changed on prod AM6 and preprod

[the remark this answers is not in the document]
Comment (ashangit) [reply readers/hdfs/src/main/java/com/criteo/hadoop/garmadon/hdfs/offset/HdfsOffsetComputer.java]: even if we could must be done in another commit

File: readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/writer/ProtoParquetWriterWithOffsetTest.java [hunk 6]
```diff
@@ -67,6 +69,38 @@ public void closeWithSomeEvents() throws IOException {
...
 
+    @Test
+    public void closeWithExistingIndexFile() throws IOException {
+        final LocalDateTime today = LocalDateTime.now();
+        final ProtoParquetWriter<Message> writerMock = mock(ProtoParquetWriter.class);
+        final Path tmpPath = new Path("tmp");
+        final Path finalPath = new Path("final");
+        final FileSystem fsMock = mock(FileSystem.class);
+        final FileStatus fileStatusMock = mock(FileStatus.class);
+        final Path path = new Path("final/" + today.format(DateTimeFormatter.ISO_DATE) + "/1.index=1.0");
```
Comment (gray-matter) [inline readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/writer/ProtoParquetWriterWithOffsetTest.java]: new Path(finalPath, ...)

File: readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/writer/ProtoParquetWriterWithOffsetTest.java [hunk 7]
```diff
@@ -67,6 +69,38 @@ public void closeWithSomeEvents() throws IOException {
...
+        final Path pathRes = new Path("final/" + today.format(DateTimeFormatter.ISO_DATE) + "/1.index=2.3");
```
Comment (gray-matter) [inline readers/hdfs/src/test/java/com/criteo/hadoop/garmadon/hdfs/writer/ProtoParquetWriterWithOffsetTest.java]: new Path(finalPath, ...)

**Pull request merged**